templates/addons/aws/modules/vpc/main.tf

```diff
@@ -15,3 +15,22 @@ module "vpc" {
   one_nat_gateway_per_az = false
   enable_dns_hostnames   = true
 }
+
+data "aws_route_tables" "private_route_table" {
+  vpc_id = module.vpc.vpc_id
+
+  filter {
+    name   = "tag:Name"
+    values = ["${var.env_namespace}-vpc-private"]
+  }
+}
+
+resource "aws_vpc_endpoint" "s3" {
```

Comment:
Does this only affect when we download/upload from/to the S3 in the application, or does it also affect when we write the AWS services' logs to S3?
Currently, our template is being used for ALB logging, so I'm not sure if we need it right now 🤔

Reply:
I might be wrong, but from my understanding: Fargate task create logs into CloudWatch. CloudWatch manages and stores the logs into S3.

If that's correct, this PR does not cover it as it would require a [VPC Endpoint Gateway for the Logs service](https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/cloudwatch-logs-and-interface-VPC.html).

This might also be useful as some issues about costs can be found — e.g. https://stackoverflow.com/questions/73940354/aws-fargate-logging-to-cloudwatch-causes-nat-traffic-how-to-avoid

Reply:
Now to the question of whether the current Gateway is useful or not: 

As mentioned in the GH issue, the ECS Registry stores the container images in S3. Deployments generate NAT Gateway when fetching the latest version of the container image from the registry (at least that's what we understood in Flexi).
Besides, this Gateway does not add any extra costs, so if that prevent future teams from spending a week to resolve the issue, it's quite a good deal ^^

Reply:
> Fargate task create logs into CloudWatch. CloudWatch manages and stores the logs into S3.
> 
> If that's correct, this PR does not cover it as it would require a [VPC Endpoint Gateway for the Logs service](https://docs.aws.amazon.com/AmazonCloudWatch/latest/logs/cloudwatch-logs-and-interface-VPC.html).

I think you understand it correctly. So in this PR, we can support the `log` service. But for the S3 service, as you mentioned, we need to add [S3 addon](https://github.com/nimblehq/infrastructure-templates/issues/152) first, then we can add the above one.

Reply:
I don't think we "need" the S3 addon.

The use cases for S3 in AWS are numerous. Even if you don't plan to use S3 at the start of a project, you might end up with some services/features that do. As that Endpoint Gateway is free, do you see any reason for not including it by default? 👀 
(as I only see the reason to have it by default: not having a budget alarm, extra costs and time consumed in the investigation ^^)

Reply:
Yes, I got your point. But the current template doesn't use S3 directly, so we might need to change the service from `s3` to `log` instead (as we are sending logs to CloudWatch, not S3 directly).
As your issue mentioned S3 file transfer, we can ship it with S3 addon later when we add it.
At the moment, we might change to support for the CloudWatch

Reply:
Okay, let me create another issue for S3 and rename this PR/issue :harold:

Reply:
> As your issue mentioned S3 file transfer, we can ship it with S3 addon later when we add it.

I don't understand why it should be in the S3 addon, maybe better to ship it with the ECS addon? Since it is directly related to ECS service, and only indirectly related to S3

Reply:
🤔 We should find a way to add it when we use both ECS and S3. Do you think we should add both `aws_vpc_endpoint` for `logs` and `s3` to the ECS module regardless of whether logs or S3 are used?

Reply:
Yep, I think they are not related to `s3`, they are related to `logs` and `ecs`. 

In my understanding `s3` module is more about setting `s3` buckets for the app needs, not for infra needs.

Reply:
Okay, so I will edit this PR to include both VPC Endpoint Gateways (one bounded to S3, the other one to the logs).

From what I understand, I should move this code from the VPC module, into the ECS module, even if VPC Endpoints are attached to a VPC (and have route tables from the subets), right? 👀

Reply:
If we shift this to the `VPC` module, could we manage a "partial" init? Say, if a user chooses the ECS module, we include a VPC endpoint, and if the user opts for the logs module, we append another VPC endpoint. I'm just worried it might complicate the implementation

Reply:
@hoangmirs @Nihisil I don't think I have the availability to handle the partial init. Could you suggest me what is best for the current work:
- Merge this PR as current first
- Close this current PR

And for the partial init. suggestion:
- Create an issue for it (assuming someone has time to handle it)?
- Let it go if we have too many other priorities.

Based on your answers I will take the needed actions.
Thanks 🙇

Reply:
I will leave decision on @hoangmirs since he has more clear view of the template architecture

Reply:
We can merge the current implementation first; then, we can have another issue to investigate a better approach for handling this 🙏

Reply:
Okay, PR marked as `Ready for review` then 😇  Thanks 🙏